Playlist Entry Limits

JTProjectFun opened the pull request:
Here's the start of my contribution to the project - playlist limiting for events.  There's a global limit setting, which as far as I can tell is correctly assigned when a new event is created.  Each event's limit can then be changed individually.  When a client reaches their request limit, no form to add another entry is shown. Similarly when a guest accesses the unique playlist URL for the event they're told they can't currently add entries.

Please check I haven't overlooked anything w.r.t. the setting storage - everything else I've tried to follow coding standards I've seen within the project.

JTProjectFun committed:
Setup for playlist limits done. Still to do - limit checking in playlist adding by clients & guests

JTProjectFun committed:
Think we're gonna need this change too, maybe I overlooked it...

JTProjectFun committed:
Added initial playlist limit functionality to client & guest playlist pages.

JTProjectFun committed:
Some indent & formatting changes

mikeyhoward1977 commented on includes/admin/events/metaboxes.php:
```diff
@@ -609,6 +609,7 @@ function mdjm_event_metabox_options_playlist_row( $event_id )	{
 	global $mdjm_event, $mdjm_event_update;
 
 	$enable_playlist = mdjm_get_option( 'enable_playlists', true );
+	$pl_limit = $mdjm_event->get_meta( '_mdjm_event_playlist_limit' );
```
Please use `$playlist_limit` rather than `$pl_limit`. Also align the variables... i.e.

    $enable_playlist = mdjm_get_option( 'enable_playlists', true );
    $playlist_limit  = $mdjm_event->get_meta( '_mdjm_event_playlist_limit' );

mikeyhoward1977 commented on includes/admin/events/metaboxes.php:
```diff
@@ -624,6 +625,14 @@ function mdjm_event_metabox_options_playlist_row( $event_id )	{
 
+    <p><?php _e ( 'Playlist Limit', 'mobile-dj-manager' ); ?>
+
+    <?php echo MDJM()->html->text( array(
```
Can this be a number field?

Again, check alignment.

    <p><?php echo MDJM()->html->text( array(
          'name'  => '_mdjm_event_playlist_limit',
          'class' => '',
          'value' => ! empty( $playlist_limit ) ? $playlist_limit : ''
    ) ); ?></p>

Commit:
Corrected some things Mike H requested be changed.

mikeyhoward1977 commented on includes/admin/mdjm.php:
```diff
@@ -99,6 +99,7 @@ public function mdjm_settings()	{
 				define( 'MDJM_SHORTDATE_FORMAT', isset( $mdjm_settings['main']['short_date_format'] ) ? $mdjm_settings['main']['short_date_format'] : 'd/m/Y' );
 				define( 'MDJM_EVENT_PREFIX', isset( $mdjm_settings['events']['event_prefix'] ) ? $mdjm_settings['events']['event_prefix'] : '' );
 				define( 'MDJM_PLAYLIST_ENABLE', !empty( $mdjm_settings['playlist']['enable_playlists'] ) ? true : false );
+                                define( 'MDJM_PLAYLIST_LIMIT', isset( $mdjm_settings['playlist']['playlist_limit'] ) ? $mdjm_settings['playlist']['playlist_limit'] : '0' );
```
This can be removed as we're slowly deprecating the use of constants as well as the entire mdjm.php file :)

mikeyhoward1977 commented on includes/admin/events/metaboxes.php:
```diff
@@ -624,6 +625,13 @@ function mdjm_event_metabox_options_playlist_row( $event_id )	{
+
+    <?php echo MDJM()->html->number( array(
+       		 	 'name'  => '_mdjm_event_playlist_limit',
+                	 'class' => '',
```
The empty class can be removed from the array.

Additionally we need to be checking if it is a new event (in which case we take the global value) or if this is an existing event (in which case we read the meta value).

Currently, when I go to an event (new or existing) the field is empty, regardless of what I have set in settings

    if ( ! $mdjm_event_update ) {
        $playlist_limit = mdjm_get_option( 'playlist_limit' );
    } else {
        $playlist_limit = mdjm_get_playlist_limit( $mdjm_event->ID );
    }

    if ( ! $playlist_limit ) {
        $playlist_limit = 0;
    }

You should then be able to change the 'value' to simply `$playlist_limit`

JTProjectFun replied:
Aha yes - I wasn't sure if I'd succeeded in creating the postmeta in code properly.  For testing I manually manipulated the database & adding new events after that seemed to work ok

JTProjectFun replied:
The latest commit might just cover everything so far. Thanks!

mikeyhoward1977 commented on includes/admin/settings/register-settings.php:
```diff
@@ -477,7 +477,17 @@ function mdjm_get_registered_settings()	{
+					'playlist_limit' => array(
+						'id'          => 'playlist_limit',
+						'name'        => __( 'Playlist Limit?', 'mobile-dj-manager' ),
+						'desc'        => sprintf( __( 'If this number is non-zero, event playlists will be limited to this number of entries ', 'mobile-dj-manager' ), mdjm_get_label_singular( true ) ),
```
1. Move the setting up to above the **Close the Playlist** setting.

2. Change the description to be more inline with others...

_Set the global limit for the number of entries a playlist can contain or enter <code>0</code> for no limit._

3. We shouldn't ever output the words 'event' or 'events' as this would mean that we would break functionality for people who are hooking into the `mdjm_default_events_name` which allows the word 'event' to be replaced with whatever term a business prefers. Instead we use `mdjm_get_label_singular()` and `mdjm_get_label_plural()` for 'event' and 'events' respectively. Default is that they will have a capital 'E'. To fix to lowercase, pass 'true' to either function.

With that in mind, your `sprintf` command should be something like...

`sprintf( __( 'If this number is non-zero, %s playlists will be limited to this number of entries ', 'mobile-dj-manager' ), mdjm_get_label_singular( true ) ),`

4. Make type 'number'

mikeyhoward1977 commented on includes/events/class-mdjm-event.php:
```diff
@@ -1079,6 +1080,14 @@ public function playlist_is_enabled() {
+	 * @return number
+	 */
+	public function get_playlist_limit() {
+		$pl_limit = get_post_meta( $this->ID, '_mdjm_event_playlist_limit', true );
```
Highly recommend only pulling this info in one place so for this method simply use the following and utilise the function you created...

    public function get_playlist_limit() {
        return mdjm_get_playlist_limit( $this->ID );
    } // get_playlist_limit

mikeyhoward1977 commented on includes/playlist/playlist-functions.php:
```diff
@@ -344,6 +344,14 @@ function mdjm_get_event_playlist_categories( $event_id, $args=array() )	{
+ */
+function mdjm_get_playlist_limit ( $event_id )  {
+     $playlist_limit = get_post_meta( $event_id, '_mdjm_event_playlist_limit', true );
+     return $playlist_limit;
```
Let's allow filtering of this value in case there is ever a need for a developer or an add-on to override the setting...

`return apply_filters( 'mdjm_event_playlist_limit', $playlist_limit, $event_id );`

See [https://developer.wordpress.org/reference/functions/apply_filters/](https://developer.wordpress.org/reference/functions/apply_filters/)

mikeyhoward1977 reviewed (changes requested):
Hey Justin - Thank you so much for contributing!

I am working through and making some suggestions for you to review. Please excuse any quick notes, I'm just trying to get through it quickly whilst providing info for you.

I have yet to review the template files. I will do that shortly. One thing I did see is that you are checking if the number of entries exceeds the limit prior to checking if the playlist is open. Consider changing this around as the open check is much quicker and if closed, performing the count and verification is simply unnecessary and time consuming

Commit:
Fix minor spacing

mikeyhoward1977 commented on includes/playlist/playlist-functions.php:
```diff
@@ -344,6 +344,14 @@ function mdjm_get_event_playlist_categories( $event_id, $args=array() )	{
+/** Get the playlist limit for this event
+ *
+ */
+function mdjm_get_playlist_limit ( $event_id )  {
```
Remove whitespace between function name and opening bracket

JTProjectFun committed:
Main change summary: Removed empty class entry in metaboxes.php. Moved playlist limit checking to inside playlist open loop.

JTProjectFun committed:
Call mdjm_get_playlist_limit instead of grabbing the event postmeta

mikeyhoward1977 commented on includes/playlist/playlist-functions.php:
```diff
@@ -344,6 +344,14 @@ function mdjm_get_event_playlist_categories( $event_id, $args=array() )	{
+ */
+function mdjm_get_playlist_limit ( $event_id ) {
+     $playlist_limit = get_post_meta( $event_id, '_mdjm_event_playlist_limit', true );
+     return $playlist_limit;
```
Please replace `return $playlist_limit;` with `return apply_filters( 'mdjm_event_playlist_limit', $playlist_limit, $event_id );` to allow for filtering of the value by add-ons and/or devs

mikeyhoward1977 commented on includes/admin/events/metaboxes.php:
```diff
@@ -624,6 +634,12 @@ function mdjm_event_metabox_options_playlist_row( $event_id )	{
+
+    <?php echo MDJM()->html->number( array(
+       			'name'  => '_mdjm_event_playlist_limit',
+ 	              	'value' => ! empty( $playlist_limit ) ? $playlist_limit : '' ) ); ?></p>
```
There is no check required here... just use `'value' => $playlist_limit`

mikeyhoward1977 commented on templates/playlist/playlist-client.php:
```diff
@@ -41,56 +41,61 @@
+                		</table>
+            	</form>
+		<?php else : ?>
+			<p><?php printf( __( 'You have used %s tracks our of your allowance of %s in the playlist for your %s.  Existing playlist entries are displayed below.', 'mobile-dj-manager' ), $entries_in_playlist,$event_playlist_limit, mdjm_get_label_singular( true ) ); ?></p>
```
We need to make use of the `_n()` function here to return the correct term for 'tracks' - i.e. singular or plural.

Example;

    <?php printf(
        __( 'You have used %d %s our of your allowance of %d in the playlist for your %s.  Existing playlist entries are displayed below.', 'mobile-dj-manager' ),
        $entries_in_playlist,
        _n( 'track', 'tracks', $entries_in_playlist, 'mobile-dj-manager' ),
        $event_playlist_limit,
        mdjm_get_label_singular( true )
    ); ?>

[See here for reference](https://developer.wordpress.org/reference/functions/_n/)

Also note correct use of '%d' for numeric values versus '%s' for strings.

mikeyhoward1977 commented on templates/playlist/playlist-client.php:
```diff
@@ -41,56 +41,61 @@
-                    </tr>
-                </table>
-            </form>
+       	<?php if( $mdjm_event->playlist_is_open() ) : ?>
```
Please ensure that each new `if` is indented correctly and that the related `else` and `endif` statements are indented the same so it is easy to read when there is a lot of code.

JTProjectFun replied:
Bah I *knew* my indenting was gonna get me into trouble.  I'm editing code via SSH on a live server. In NANO.  Ha!

mikeyhoward1977 reviewed (changes requested):
I think you're almost there :)

Once you have made these changes I'll run some further tests and look to merge the pull request

JTProjectFun committed:
changed return value of mdjm_get_playlist_limit to allow for filtering

JTProjectFun committed:
Fixed indenting?

JTProjectFun committed:
Moved limit detection to inside 'is playlist open' loop in guest playlist template.

JTProjectFun committed:
Changed field value to be just the playlist_limit value. No need for a check here

mikeyhoward1977 reviewed (changes requested):
The syntax is incorrect here. You are missing the close brackets for `number(` and `array(`

Commit:
Fixed syntax error by adding missing close brackets to array & number

Commit:
Set correct playlist limit

If a new event OR event status is unattended, read from the global
setting. Otherwise read from the event meta key

Commit:
Alignment and whitespace

Commit:
Added function comments

All functions must adhere to the standards and contain comments that
contain the version at which they were intriduced, the parameters being
passed and also the return value

Commit:
Renamed function

Inserted 'event' into function name as the function is pulling data
directly from an event meta key

Commit:
Added function to retrieve global playlist limit

Commit:
Toggle visibility of event playlist limit setting

If the playlist for an event is not enabled the limit setting is
redundant so let’s use JS to toggle its display ensuring it is only
visible when required

Commit:
Label needed a colon
